File: backend/app/einvoice/models/einvoice.py
"""What the government portal gave back for an invoice, and for its movement.

Two registrations, one module, because they share one portal, one set of
credentials and one failure story: an e-invoice is registered with the Invoice
Registration Portal and an e-way bill is raised from that same registration
for the goods it covers.

**Every registration records the mode it was made in.** A sandbox registration
is a rehearsal -- no return was filed, no IRN exists at the authority, and the
number on it means nothing outside this database. A row that could not say
which it was would be a document somebody eventually presents at a check post.
So `mode` is NOT NULL, there is no default that could quietly become LIVE, and
the reference the sandbox mints is prefixed so it cannot be mistaken even out
of context.
"""

from datetime import date, datetime
from decimal import Decimal
from enum import StrEnum
from uuid import UUID

from sqlalchemy import (
    JSON,
    Date,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database.entity import BaseEntity
from app.core.database.types import UTCDateTime, UUIDType


class RegistrationMode(StrEnum):
    """Whether a registration was filed or rehearsed.

    There is no third value and no default. A firm switches to LIVE
    deliberately, with credentials, and every row written before that says so
    for ever after.
    """

    SANDBOX = "SANDBOX"
    LIVE = "LIVE"


class RegistrationStatus(StrEnum):
    """Where a registration got to.

    FAILED is a real state rather than an absence: the portal refused, it said
    why, and that answer is worth keeping. Retrying writes over the error on
    the same row, because one invoice has one registration.
    """

    PENDING = "PENDING"
    REGISTERED = "REGISTERED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class EInvoiceRegistration(BaseEntity):
    """One sales invoice, as the Invoice Registration Portal knows it."""

    __tablename__ = "einvoice_registrations"
    __table_args__ = (
        # One invoice, one registration. A second would leave two IRNs for one
        # supply and nothing to say which the customer holds.
        UniqueConstraint(
            "firm_id",
            "sales_invoice_id",
            name="UQ_einvoice_registrations_invoice",
        ),
        Index("IX_einvoice_registrations_firm_status", "firm_id", "status"),
    )

    #: No foreign key: `firms` lives only in the platform schema.
    firm_id: Mapped[UUID] = mapped_column(UUIDType(), nullable=False, index=True)
    sales_invoice_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("sales_invoices.id", ondelete="RESTRICT"),
        nullable=False,
    )
    mode: Mapped[str] = mapped_column(String(20), nullable=False)
    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default=RegistrationStatus.PENDING.value,
        server_default=RegistrationStatus.PENDING.value,
    )
    #: The 64-character hash the portal returns. Sandbox mints one prefixed
    #: `SBX` so it is recognisable on its own, away from this row.
    irn: Mapped[str | None] = mapped_column(String(80))
    acknowledgement_number: Mapped[str | None] = mapped_column(String(40))
    acknowledged_at: Mapped[datetime | None] = mapped_column(UTCDateTime)
    #: The signed QR the customer's copy has to carry. Held as text because it
    #: is a JWT, not a picture -- rendering is the printer's job.
    signed_qr_code: Mapped[str | None] = mapped_column(Text)
    signed_invoice: Mapped[str | None] = mapped_column(Text)
    #: What the portal said when it refused. Kept on the row rather than only
    #: in a log, because the person who has to fix the invoice is looking at
    #: the invoice.
    error_code: Mapped[str | None] = mapped_column(String(40))
    error_message: Mapped[str | None] = mapped_column(Text)
    attempts: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default="0"
    )
    cancelled_at: Mapped[datetime | None] = mapped_column(UTCDateTime)
    cancellation_reason: Mapped[str | None] = mapped_column(String(200))
    #: Exactly what was sent, so a refusal can be argued with. The portal
    #: rejects on the payload it received, not on the document as it looks
    #: today.
    request_payload: Mapped[dict[str, object] | None] = mapped_column(JSON)


class EWayBillStatus(StrEnum):
    """Where an e-way bill got to."""

    PENDING = "PENDING"
    GENERATED = "GENERATED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class TransportMode(StrEnum):
    """How the goods are moving.

    The portal takes a code; these are the names a person uses. Road is the
    only one that requires a vehicle number, which is why the service asks for
    one only then.
    """

    ROAD = "ROAD"
    RAIL = "RAIL"
    AIR = "AIR"
    SHIP = "SHIP"


class EWayBill(BaseEntity):
    """One consignment's e-way bill, raised against an invoice."""

    __tablename__ = "eway_bills"
    __table_args__ = (
        UniqueConstraint("firm_id", "sales_invoice_id", name="UQ_eway_bills_invoice"),
        Index("IX_eway_bills_firm_status", "firm_id", "status"),
    )

    firm_id: Mapped[UUID] = mapped_column(UUIDType(), nullable=False, index=True)
    sales_invoice_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("sales_invoices.id", ondelete="RESTRICT"),
        nullable=False,
    )
    mode: Mapped[str] = mapped_column(String(20), nullable=False)
    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default=EWayBillStatus.PENDING.value,
        server_default=EWayBillStatus.PENDING.value,
    )
    eway_bill_number: Mapped[str | None] = mapped_column(String(40))
    #: When the bill stops being valid. The portal decides it from the
    #: distance, so it is stored rather than computed -- a locally computed
    #: expiry that disagreed with the authority's is worse than none.
    valid_until: Mapped[date | None] = mapped_column(Date)
    distance_km: Mapped[Decimal] = mapped_column(
        Numeric(9, 2), nullable=False, default=Decimal("0"), server_default="0"
    )
    transport_mode: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default=TransportMode.ROAD.value,
        server_default=TransportMode.ROAD.value,
    )
    transporter_id: Mapped[str | None] = mapped_column(String(40))
    transporter_name: Mapped[str | None] = mapped_column(String(200))
    vehicle_number: Mapped[str | None] = mapped_column(String(20))
    error_code: Mapped[str | None] = mapped_column(String(40))
    error_message: Mapped[str | None] = mapped_column(Text)
    cancelled_at: Mapped[datetime | None] = mapped_column(UTCDateTime)
    cancellation_reason: Mapped[str | None] = mapped_column(String(200))
    request_payload: Mapped[dict[str, object] | None] = mapped_column(JSON)

File: desktop/lib/models/einvoice.dart
import 'entities.dart';

/// What the tax authority knows about one invoice.
///
/// [mode] is never absent. A SANDBOX registration is a rehearsal: nothing was
/// filed, and the reference means nothing outside this database. Every screen
/// that shows a reference has to show the mode beside it, or somebody
/// eventually presents a rehearsal at a check post.
class EInvoiceRegistrationRecord {
  const EInvoiceRegistrationRecord({
    required this.id,
    required this.salesInvoiceId,
    required this.mode,
    required this.status,
    this.irn = '',
    this.acknowledgementNumber = '',
    this.signedQrCode = '',
    this.errorCode = '',
    this.errorMessage = '',
    this.attempts = 0,
    this.cancellationReason = '',
  });

  final String id;
  final String salesInvoiceId;

  /// SANDBOX or LIVE.
  final String mode;
  final String status;
  final String irn;
  final String acknowledgementNumber;
  final String signedQrCode;
  final String errorCode;
  final String errorMessage;
  final int attempts;
  final String cancellationReason;

  bool get isRegistered => status == 'REGISTERED';
  bool get isSandbox => mode == 'SANDBOX';

  /// What to show where the reference goes, mode included.
  String get referenceLabel {
    if (!isRegistered) return status;
    return isSandbox ? '$irn  (sandbox — nothing filed)' : irn;
  }

  factory EInvoiceRegistrationRecord.fromJson(Json json) =>
      EInvoiceRegistrationRecord(
        id: stringValue(json['id']),
        salesInvoiceId: stringValue(json['sales_invoice_id']),
        mode: stringValue(json['mode']),
        status: stringValue(json['status']),
        irn: stringValue(json['irn']),
        acknowledgementNumber: stringValue(json['acknowledgement_number']),
        signedQrCode: stringValue(json['signed_qr_code']),
        errorCode: stringValue(json['error_code']),
        errorMessage: stringValue(json['error_message']),
        attempts: (json['attempts'] as num?)?.toInt() ?? 0,
        cancellationReason: stringValue(json['cancellation_reason']),
      );
}

/// What the authority knows about one consignment.
class EWayBillRecord {
  const EWayBillRecord({
    required this.id,
    required this.salesInvoiceId,
    required this.mode,
    required this.status,
    this.ewayBillNumber = '',
    this.validUntil = '',
    this.distanceKm = '0',
    this.transportMode = 'ROAD',
    this.transporterId = '',
    this.transporterName = '',
    this.vehicleNumber = '',
    this.errorCode = '',
    this.errorMessage = '',
  });

  final String id;
  final String salesInvoiceId;
  final String mode;
  final String status;
  final String ewayBillNumber;

  /// When the bill stops being valid. The authority decides it from the
  /// distance, so it is shown as given rather than recomputed here.
  final String validUntil;
  final String distanceKm;
  final String transportMode;
  final String transporterId;
  final String transporterName;
  final String vehicleNumber;
  final String errorCode;
  final String errorMessage;

  bool get isGenerated => status == 'GENERATED';
  bool get isSandbox => mode == 'SANDBOX';

  String get referenceLabel {
    if (!isGenerated) return status;
    final String suffix = isSandbox ? '  (sandbox — nothing filed)' : '';
    return validUntil.isEmpty
        ? '$ewayBillNumber$suffix'
        : '$ewayBillNumber  ·  valid to $validUntil$suffix';
  }

  factory EWayBillRecord.fromJson(Json json) => EWayBillRecord(
        id: stringValue(json['id']),
        salesInvoiceId: stringValue(json['sales_invoice_id']),
        mode: stringValue(json['mode']),
        status: stringValue(json['status']),
        ewayBillNumber: stringValue(json['eway_bill_number']),
        validUntil: stringValue(json['valid_until']),
        distanceKm: stringValue(json['distance_km']),
        transportMode: stringValue(json['transport_mode']),
        transporterId: stringValue(json['transporter_id']),
        transporterName: stringValue(json['transporter_name']),
        vehicleNumber: stringValue(json['vehicle_number']),
        errorCode: stringValue(json['error_code']),
        errorMessage: stringValue(json['error_message']),
      );
}
